chalk-solve/src/infer/unify.rs

```diff
@@ -112,16 +115,41 @@ impl<'t, I: Interner> Unifier<'t, I> {
                     .expect("unification of two unbound variables cannot fail"))
             }
 
-            // Unifying an inference variables with a non-inference variable.
-            (&TyData::InferenceVar(var), &TyData::Apply(_))
-            | (&TyData::InferenceVar(var), &TyData::Placeholder(_))
-            | (&TyData::InferenceVar(var), &TyData::Dyn(_))
-            | (&TyData::InferenceVar(var), &TyData::Function(_)) => self.unify_var_ty(var, b),
+            // Tried to unify mis-matching inference variables (not caught by prior match arm)
```

Comment:
Unifying a "general" type variable with an "int/float" type variable should be ok. In rustc, I believe we just "bind" the general one to the int/float one.

Reply:
Without this change, the SLG solver passes the tests and the recursive solver fails the `T = N, N = usize` test (and possibly more). Perhaps this change isn't needed (just a fix for the recursive solver), or is the SLG solver just lucky with the order it chooses to unify the types?

After adding the change the recursive solver fails on `T = N, T = char` because individually `?0` can unify with both `?1i` and `char`. That is, the SLG solver unifies `?0` with `char` first, then `?0=char` with `?1i` (fails here). The recursive solver unifies `?0` with `?1i` and then `?0` with `char` (succeeds, because the `?1i` kind data was lost or not used).

I'm unsure of how to "bind" the general inference variable to the more specific integer/float variable, since `unify_var_var` is in `ena` and only knows about `InferenceVar`. I suspect this would fix the problem because then the recursive solver would try to unify `?0=?1i` with `char` and fail.

Reply:
I think I figured it out. I didn't think that `unify_var_value` with a `Ty::InferenceVar` as the value would work, but it does.